modules/tinymce/src/core/main/ts/html/ExtractInvalidNode.ts

```diff
@@ -0,0 +1,152 @@
+/**
+ * Copyright (c) Tiny Technologies, Inc. All rights reserved.
+ * Licensed under the LGPL or a commercial license.
+ * For LGPL see License.txt in the project root for license information.
+ * For commercial licenses see https://www.tiny.cloud/
+ */
+
+import { Fun } from '@ephox/katamari';
+
+import AstNode from '../api/html/Node';
+import Schema from '../api/html/Schema';
+import Tools from '../api/util/Tools';
+import { hasOnlyChild, isEmpty } from './ParserUtils';
+
+const removeOrUnwrapInvalidNode = (node: AstNode, schema: Schema, originalNodeParent: AstNode = node.parent) => {
+  if (schema.getSpecialElements()[node.name]) {
+    node.empty().remove();
+  } else {
+    // are the children of `node` valid children of the top level parent?
+    // if not, remove or unwrap them too
+    const children = node.children();
+    for (const childNode of children) {
+      if (!schema.isValidChild(originalNodeParent.name, childNode.name)) {
+        removeOrUnwrapInvalidNode(childNode, schema, originalNodeParent);
+      }
+    }
+    node.unwrap();
+  }
+};
+
+const extractInvalidNode = (node: AstNode, schema: Schema, fixed: Set<AstNode> = new Set(), onCreate: (newNode: AstNode) => void = Fun.noop) => {
```

Comment:
Return types here please
```suggestion
const extractInvalidNode = (node: AstNode, schema: Schema, fixed: Set<AstNode> = new Set(), onCreate: (newNode: AstNode) => void = Fun.noop): void => {
```